security/pkg/nodeagent/sds/sds_stream_test.go

```diff
@@ -0,0 +1,394 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package sds
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	sds "github.com/envoyproxy/go-control-plane/envoy/service/secret/v3"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"k8s.io/apimachinery/pkg/util/uuid"
+
+	"istio.io/istio/security/pkg/nodeagent/cache"
+	"istio.io/istio/security/pkg/nodeagent/model"
+	"istio.io/istio/security/pkg/nodeagent/util"
+)
+
+const (
+	ValidProxyID        = "sidecar~127.0.0.1~SecretsPushStreamOne~local"
+	InValidProxyID      = "invalid~sidecar~127.0.0.1~SecretsPushStreamOne~local"
+	BlockGenSecretError = "BLOCK_GENERATE_SECRET_FOR_TEST_ERROR"
+)
+
+/*
+	Validate that secrets are cleaned after connection is closed in such two cases:
+	1. workflow are run completely and connection is closed
+	2. workflow are terminated and connection is closed  after secret are generated
+*/
+func TestSDSAgentStreamWithCacheAndConnectionCleaned(t *testing.T) {
+	setup := StartStreamTest(t)
+	defer setup.server.Stop()
+
+	conn, stream := createSDSStream(t, setup.socket, fakeToken1)
+	notifyChan := make(chan notifyMsg)
+
+	go testSDSSuccessIngressStreamCache(stream, ValidProxyID, notifyChan)
+	// verify that the first SDS request sent by two streams do not hit cache.
+	waitForStreamSecretCacheCheck(t, setup.secretStore, false, 1)
+	waitForStreamNotificationToProceed(t, notifyChan, "notify push secret 1")
+	secretKeyMap := make(map[interface{}]bool)
+	setup.secretStore.secrets.Range(func(key, value interface{}) bool {
+		secretKeyMap[key] = true
+		return false
+	})
+	conn.Close()
+	// verify the cache is cleaned when connection is closed
+	waitForSecretCacheCleanUp(t, setup.secretStore, secretKeyMap)
+
+	conn, stream = createSDSStream(t, setup.socket, fakeToken1)
+	go testSDSTerminatedIngressStreamCache(stream, InValidProxyID, notifyChan)
```

Comment:
Please add a comment above this go testSDSTerminatedIngressStreamCache(): When proxy ID has "invalid", SDS server closes the connection and returns an error.